otter/test/convergence/test_service.py

```diff
@@ -1,95 +1,409 @@
+                            version=self.version),
+                 lambda i: self.deletions.append(True))
+            ]),
+            _get_dispatcher(),
+        ])
+
+    def test_success(self):
+        """
+        runs execute_convergence and returns None, then deletes the dirty flag.
+        """
+        calls = []
+
+        def execute_convergence(tenant_id, group_id, log):
+            return Effect(Func(
+                lambda: calls.append((tenant_id, group_id, log))))
+
+        eff = converge_one(
+            self.log, make_lock_set(), self.tenant_id, self.group_id,
+            self.version,
+            execute_convergence=execute_convergence)
+        result = sync_perform(self.dispatcher, eff)
+        self.assertEqual(result, None)
+        self.assertEqual(
+            calls,
+            [(self.tenant_id, self.group_id,
+              matches(IsBoundWith(tenant_id=self.tenant_id,
+                                  group_id=self.group_id)))])
+        self.assertEqual(self.deletions, [True])
+        self.log.msg.assert_any_call(
+            'mark-clean-success',
+            tenant_id=self.tenant_id, group_id=self.group_id)
+
+    def test_non_concurrent(self):
+        """Runs execute_convergence non-concurrently, based on the group ID."""
+        calls = []
+
+        def execute_convergence(tenant_id, group_id, log):
+            return Effect(Func(lambda: calls.append('should not be run')))
+
+        lock_set = partial(non_concurrently, Reference(pset([self.group_id])))
+        eff = converge_one(
+            self.log, lock_set, self.tenant_id,
+            self.group_id, self.version,
+            execute_convergence=execute_convergence)
+        result = sync_perform(self.dispatcher, eff)
+        self.assertEqual(result, None)
+        self.assertEqual(calls, [])
+
+    def test_no_scaling_group(self):
+        """
+        When the scaling group disappears, a fatal error is logged and the
+        dirty flag is cleaned up.
+        """
+        def execute_convergence(tenant_id, group_id, log):
+            return Effect(Error(NoSuchScalingGroupError(tenant_id, group_id)))
+
+        eff = converge_one(
+            self.log, make_lock_set(), self.tenant_id, self.group_id,
+            self.version,
+            execute_convergence=execute_convergence)
+        result = sync_perform(self.dispatcher, eff)
+        self.assertEqual(result, None)
+
+        self.log.err.assert_any_call(
+            CheckFailureValue(NoSuchScalingGroupError(self.tenant_id,
+                                                      self.group_id)),
+            'converge-fatal-error',
+            tenant_id=self.tenant_id, group_id=self.group_id)
+
+        self.assertEqual(self.deletions, [True])
+        self.log.msg.assert_any_call(
+            'mark-clean-success',
+            tenant_id=self.tenant_id, group_id=self.group_id)
+
+    def test_concurrent_error_ignored(self):
```

Comment:
As @radix points out, #3/#4 are invalid. #3 is poorly described and a duplicate of #2, and the purpose of them was mostly to demonstrate that a call coming in the middle of another one will not interfere with the original call.  which isn't really necessary.